### PR TITLE
feat: add NewTicker to clock testing library

### PR DESCRIPTION
**Pull request by spikecurtis**
Adds `NewTicker` and the corresponding `Ticker` struct, for matching the standard library or for tickers that need to be part of more complex `select` statements such that `TickerFunc` doesn't work.

### CODE REVIEW

**Comment by spikecurtis** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/coder/coder/13593?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#13593** <a href="https://app.graphite.dev/github/pr/coder/coder/13593?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#13592** <a href="https://app.graphite.dev/github/pr/coder/coder/13592?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @spikecurtis and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: clock/clock.go [hunk 1]
```diff
@@ -10,6 +10,12 @@ import (
...
 type Clock interface {
+	// NewTicker returns a new Ticker containing a channel that will send the current time on the
+	// channel after each tick. The period of the ticks is specified by the duration argument. The
+	// ticker will adjust the time interval or drop ticks to make up for slow receivers. The
+	// duration d must be greater than zero; if not, NewTicker will panic. Stop the ticker to
+	// release associated resources.
+	NewTicker(d time.Duration, tags ...string) *Ticker
 	// TickerFunc is a convenience function that calls f on the interval d until either the given
 	// context expires or f returns an error.  Callers may call Wait() on the returned Waiter to
 	// wait until this happens and obtain the error.
```
Comment (johnstcn) [inline clock/clock.go:21]: Suggestion: add the provisio regarding negative duration to the docs for `TickerFunc`

File: clock/mock.go [hunk 2]
```diff
@@ -67,14 +89,17 @@ func (m *Mock) NewTimer(d time.Duration, tags ...string) *Timer {
...
 		mock: m,
 	}
-	m.addTimerLocked(t)
+	if d <= 0 {
+		// zero or negative duration timer means we should immediately fire
+		// it, rather than add it.
+		go t.fire(t.mock.cur)
+		return t
+	}
+	m.addEventLocked(t)
```
Comment (johnstcn) [inline clock/mock.go:98]: nit: related to this PR, or did it sneak in from elsewhere? The behaviour does seem valid from what I can tell. https://go.dev/play/p/xYggZfm7Szw

**Review by johnstcn — APPROVED** [review]

File: clock/mock.go (hunk 2 above)
Comment (spikecurtis) [reply clock/mock.go]: It shows up when downstack PRs get merged but this one hasn't yet been restacked

**Comment by spikecurtis** [comment]
### Merge activity

* **Jun 20, 2:16 AM EDT**: @spikecurtis merged this pull request with [Graphite](https://app.graphite.dev/github/pr/coder/coder/13593).

**Pull request merged**